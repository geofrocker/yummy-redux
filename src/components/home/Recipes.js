import React from 'react';
import Since from 'react-since';

const Recipes = props =>
  (<div className="jumbotron">
    <h3>{props.title}</h3>
    <em>Added by <span className="fa fa-user" /> { props.author } about <span className="fa fa-calendar" /> <Since date={props.create_date} /> | Category: { props.category_rel}
      <a href={`/recipe/${props.recipe_id}`} id="review" className="btn btn-primary pull-right">Review</a>
    </em>

    <div>
      <h3>Ingredients</h3>
      <small>{props.ingredients}</small>
      <h3>Steps</h3>
      <small>{props.steps}</small><br /><br />
      <div className="btn-group" id="btn-group">
        <button type="button" className="btn btn-default btn-xs"><span className="fa fa-comment-o" /> Reviews:{props.reviews}</button>
        <button type="button" className="btn btn-default btn-xs"><span className="fa fa-thumbs-o-up" /> Upvotes:{props.upvotes}</button>
      </div>
    </div>
   </div>);
export default Recipes;
