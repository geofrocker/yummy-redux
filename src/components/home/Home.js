import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux'
import {bindActionCreators} from 'redux'
import * as homeRecipesActions from '../../actions/homeRecipesActions'
import Recipes from './Recipes'
import url from '../../config'
class Home extends Component {
    constructor(props, context){
        super(props);
        this.state = {
            data:[],
            showMessage:false, 
            q:'', 
            page:1, 
            has_next:false, 
            next_page:'',
            previous_page:'', 
            disablePrevious:'', 
            disableNext:'',
            url: url+'?page=1',
            pages:null
        };
        this.props.actions.loadPublicRecipes(this.state.url)
    }
    
    //Load recipes before the component recieve props
    componentWillReceiveProps(nextProps){
        this.setState({
            data:nextProps.data.Recipe_list,
            showMessage:false,
            pages: nextProps.data.total_pages
        })
        if(nextProps.data.previous_page === 'Null'){
            this.setState({
                disablePrevious: 'page-item disabled',
                previous_page: ''
                });
        }else{
            this.setState({
                previous_page: nextProps.data.previous_page,
                disablePrevious: 'page-item'
                });
        }
        if(nextProps.data.next_page === 'Null'){
            this.setState({
                disableNext: 'page-item disabled',
                next_page: '',
                });
        }else{
            this.setState({
                next_page: nextProps.data.next_page,
                disableNext: 'page-item'
                });
        }
    }

    //Handle the link to the next page from pagination
    nextPage() {
        if (this.state.next_page)
            this.props.actions.loadPublicRecipes(this.state.next_page)
            return this.setState({
                page:this.state.next_page[this.state.next_page.length -1]
            })
    }

    // Handle the link to the previous page from pagination
    previousPage() {
        if(this.state.previous_page)
        this.props.actions.loadPublicRecipes(this.state.previous_page)
        return this.setState({
            page:this.state.previous_page[this.state.previous_page.length -1]
        })
    };

    // Handle the search functionality
    handleSearch = (event) => {
        event.preventDefault();
        this.setState({
            q: event.target.value,
            page:1
        });
        if(this.state.q){
            let localurl = url +'?q='+this.state.q+'&page='+this.state.page
            this.props.actions.loadPublicRecipes(localurl)
        }
    }
    render(){
        const {data, disableNext, disablePrevious, page, pages, showMessage, }=this.state
        let loadPagination;
        if (data.length>0) {
            loadPagination =
           <div className="col-xs-11 col-sm-3 pull-right">
                <ul className="pagination">
                    <li className={disablePrevious}>
                    <a className="page-link" onClick={() => this.previousPage()}>Previous</a>
                    </li>
                    <li className={disableNext}>
                    <a className="page-link" onClick={() => this.nextPage()}>Next</a>
                    <a className="page-link">Showing {page} of {pages}</a>
                    </li>
                    
                </ul>
            </div> 
        }
        let loadNavBarContent;
        if (showMessage) {
           loadNavBarContent =
           <div className="Home jumbotron" id="search-bar">
                <h3>No recipes matched your query.</h3>
            </div>
        }
        return (
            <div className="Home">
                <div className="col-xs-12 col-sm-6 pull-right">
                    <div className="input-group mb-2 mb-sm-0">
                        <div className="input-group-addon">Search</div>
                        <input type="text" className="form-control" onChange={this.handleSearch} onKeyUp={this.handleSearch} id="search" placeholder="Enter your search key words here!"/>
                    </div>
                </div>
                <div className="col-xs-12">
                {loadNavBarContent}
                <hr/>
                    {data.map(inf =>
                    <Recipes key={inf.recipe_id} {...inf}/>
                    )}
                </div>
                {loadPagination}
            </div>
    );
    }
}

Home.propTypes = {
    data: PropTypes.object.isRequired,
  }
  
  function mapStateToProps(state, ownProps){
    return {
      data:state.publicRecipes
    }
  }
  
  function mapDispatchToProps(dispatch){
    return{
      actions: bindActionCreators(homeRecipesActions, dispatch)
    }
  }
  export default connect(mapStateToProps, mapDispatchToProps) (Home);