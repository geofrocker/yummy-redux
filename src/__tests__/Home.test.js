import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { shallow, mount } from 'enzyme';
import { Home } from '../components/home/Home';
import toJson from 'enzyme-to-json';

describe('the home component', () => {
  const props = {
    history: {
      push: () => {},
    },
    recipes: [{ recipe_id: 1, title: 'Lorem' }],
    loadRecipes: () => Promise.resolve(),
  };
  const wrapper = mount(<Home {...props} />);

  it('renders the Home class', () => {
    expect(wrapper.find('.Home')).toHaveLength(1);
  });

  it('renders the jumbotron class', () => {
    expect(wrapper.find('.col-xs-12')).toHaveLength(2);
  });
  it('renders five div jsx elements', () => {
    expect(wrapper.find('div')).toHaveLength(11);
  });
  it('renders search bar', () => {
    expect(wrapper.find('#search')).toHaveLength(1);
  });
  it('renders pagination', () => {
    wrapper.setState({ data: [{ a: 'general' }] });
    expect(wrapper.find('.pagination')).toHaveLength(1);
    expect(wrapper.find('.page-link')).toHaveLength(3);
  });
  it('renders search bar', () => {
    wrapper.setState({ showMessage: true });
    expect(wrapper.find('#search-bar')).toHaveLength(1);
  });
  it('renders search input', () => {
    expect(wrapper.find('#search')).toHaveLength(1);
  });
  it('renders Recipe class', () => {
    wrapper.setState({ data: [{ a: 'general' }] });
    expect(wrapper.find('Recipes')).toHaveLength(1);
  });

  it(' handle going back to previous page ', () => {
    wrapper.find('#prevPage').simulate('click');
  });

  it(' handle going back to next page ', () => {
    wrapper.find('#nextPage').simulate('click');
  });

  it(' handles search feature ', () => {
    wrapper.find('#search').simulate('click');
  });

  it(' handle going to the review page ', () => {
    wrapper.find('#review').simulate('click');
  });
});
