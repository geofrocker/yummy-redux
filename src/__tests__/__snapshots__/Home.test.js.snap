// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the home component  handle going back to next page  1`] = `
<Home
  history={
    Object {
      "push": [Function],
    }
  }
  loadRecipes={[Function]}
  recipes={
    Array [
      Object {
        "recipe_id": 1,
        "title": "Lorem",
      },
    ]
  }
>
  <div
    className="Home"
  >
    <div
      className="col-xs-12 col-sm-6 pull-right"
    >
      <div
        className="input-group mb-2 mb-sm-0"
      >
        <div
          className="input-group-addon"
        >
          Search
        </div>
        <input
          className="form-control"
          id="search"
          onChange={[Function]}
          onKeyUp={[Function]}
          placeholder="Enter your search key words here!"
          type="text"
        />
      </div>
    </div>
    <div>
      <div
        className="col-xs-12"
      >
        <div
          className="Home jumbotron"
          id="search-bar"
        >
          <h3>
            No recipes matched your query.
          </h3>
        </div>
        <hr />
        <Recipes
          key="1"
          recipe_id={1}
          title="Lorem"
        >
          <div
            className="jumbotron"
          >
            <h3>
              Lorem
            </h3>
            <em>
              Added by 
              <span
                className="fa fa-user"
              />
               
               about 
              <span
                className="fa fa-calendar"
              />
               
              <Since
                live={true}
              />
               | Category: 
              <a
                className="btn btn-primary pull-right"
                href="/recipe/1"
                id="review"
              >
                Review
              </a>
            </em>
            <div>
              <h3>
                Ingredients
              </h3>
              <small />
              <h3>
                Steps
              </h3>
              <small />
              <br />
              <br />
              <div
                className="btn-group"
                id="btn-group"
              >
                <button
                  className="btn btn-default btn-xs"
                  type="button"
                >
                  <span
                    className="fa fa-comment-o"
                  />
                   Reviews:
                </button>
                <button
                  className="btn btn-default btn-xs"
                  type="button"
                >
                  <span
                    className="fa fa-thumbs-o-up"
                  />
                   Upvotes:
                </button>
              </div>
            </div>
          </div>
        </Recipes>
      </div>
      <div
        className="col-xs-11 col-sm-3 pull-right"
      >
        <ul
          className="pagination"
        >
          <li>
            <a
              className="page-link"
              id="prevPage"
              onClick={[Function]}
            >
              Previous
            </a>
          </li>
          <li>
            <a
              className="page-link"
              id="nextPage"
              onClick={[Function]}
            >
              Next
            </a>
            <a
              className="page-link"
            >
              Showing 
               of 
            </a>
          </li>
        </ul>
      </div>
    </div>
    <div />
  </div>
</Home>
`;

exports[`the home component  handle going back to previous page  1`] = `
<Home
  history={
    Object {
      "push": [Function],
    }
  }
  loadRecipes={[Function]}
  recipes={
    Array [
      Object {
        "recipe_id": 1,
        "title": "Lorem",
      },
    ]
  }
>
  <div
    className="Home"
  >
    <div
      className="col-xs-12 col-sm-6 pull-right"
    >
      <div
        className="input-group mb-2 mb-sm-0"
      >
        <div
          className="input-group-addon"
        >
          Search
        </div>
        <input
          className="form-control"
          id="search"
          onChange={[Function]}
          onKeyUp={[Function]}
          placeholder="Enter your search key words here!"
          type="text"
        />
      </div>
    </div>
    <div>
      <div
        className="col-xs-12"
      >
        <div
          className="Home jumbotron"
          id="search-bar"
        >
          <h3>
            No recipes matched your query.
          </h3>
        </div>
        <hr />
        <Recipes
          key="1"
          recipe_id={1}
          title="Lorem"
        >
          <div
            className="jumbotron"
          >
            <h3>
              Lorem
            </h3>
            <em>
              Added by 
              <span
                className="fa fa-user"
              />
               
               about 
              <span
                className="fa fa-calendar"
              />
               
              <Since
                live={true}
              />
               | Category: 
              <a
                className="btn btn-primary pull-right"
                href="/recipe/1"
                id="review"
              >
                Review
              </a>
            </em>
            <div>
              <h3>
                Ingredients
              </h3>
              <small />
              <h3>
                Steps
              </h3>
              <small />
              <br />
              <br />
              <div
                className="btn-group"
                id="btn-group"
              >
                <button
                  className="btn btn-default btn-xs"
                  type="button"
                >
                  <span
                    className="fa fa-comment-o"
                  />
                   Reviews:
                </button>
                <button
                  className="btn btn-default btn-xs"
                  type="button"
                >
                  <span
                    className="fa fa-thumbs-o-up"
                  />
                   Upvotes:
                </button>
              </div>
            </div>
          </div>
        </Recipes>
      </div>
      <div
        className="col-xs-11 col-sm-3 pull-right"
      >
        <ul
          className="pagination"
        >
          <li>
            <a
              className="page-link"
              id="prevPage"
              onClick={[Function]}
            >
              Previous
            </a>
          </li>
          <li>
            <a
              className="page-link"
              id="nextPage"
              onClick={[Function]}
            >
              Next
            </a>
            <a
              className="page-link"
            >
              Showing 
               of 
            </a>
          </li>
        </ul>
      </div>
    </div>
    <div />
  </div>
</Home>
`;

exports[`the home component  handle going to the review page  1`] = `
<Home
  history={
    Object {
      "push": [Function],
    }
  }
  loadRecipes={[Function]}
  recipes={
    Array [
      Object {
        "recipe_id": 1,
        "title": "Lorem",
      },
    ]
  }
>
  <div
    className="Home"
  >
    <div
      className="col-xs-12 col-sm-6 pull-right"
    >
      <div
        className="input-group mb-2 mb-sm-0"
      >
        <div
          className="input-group-addon"
        >
          Search
        </div>
        <input
          className="form-control"
          id="search"
          onChange={[Function]}
          onKeyUp={[Function]}
          placeholder="Enter your search key words here!"
          type="text"
        />
      </div>
    </div>
    <div>
      <div
        className="col-xs-12"
      >
        <div
          className="Home jumbotron"
          id="search-bar"
        >
          <h3>
            No recipes matched your query.
          </h3>
        </div>
        <hr />
        <Recipes
          key="1"
          recipe_id={1}
          title="Lorem"
        >
          <div
            className="jumbotron"
          >
            <h3>
              Lorem
            </h3>
            <em>
              Added by 
              <span
                className="fa fa-user"
              />
               
               about 
              <span
                className="fa fa-calendar"
              />
               
              <Since
                live={true}
              />
               | Category: 
              <a
                className="btn btn-primary pull-right"
                href="/recipe/1"
                id="review"
              >
                Review
              </a>
            </em>
            <div>
              <h3>
                Ingredients
              </h3>
              <small />
              <h3>
                Steps
              </h3>
              <small />
              <br />
              <br />
              <div
                className="btn-group"
                id="btn-group"
              >
                <button
                  className="btn btn-default btn-xs"
                  type="button"
                >
                  <span
                    className="fa fa-comment-o"
                  />
                   Reviews:
                </button>
                <button
                  className="btn btn-default btn-xs"
                  type="button"
                >
                  <span
                    className="fa fa-thumbs-o-up"
                  />
                   Upvotes:
                </button>
              </div>
            </div>
          </div>
        </Recipes>
      </div>
      <div
        className="col-xs-11 col-sm-3 pull-right"
      >
        <ul
          className="pagination"
        >
          <li>
            <a
              className="page-link"
              id="prevPage"
              onClick={[Function]}
            >
              Previous
            </a>
          </li>
          <li>
            <a
              className="page-link"
              id="nextPage"
              onClick={[Function]}
            >
              Next
            </a>
            <a
              className="page-link"
            >
              Showing 
               of 
            </a>
          </li>
        </ul>
      </div>
    </div>
    <div />
  </div>
</Home>
`;

exports[`the home component  handles search feature  1`] = `
<Home
  history={
    Object {
      "push": [Function],
    }
  }
  loadRecipes={[Function]}
  recipes={
    Array [
      Object {
        "recipe_id": 1,
        "title": "Lorem",
      },
    ]
  }
>
  <div
    className="Home"
  >
    <div
      className="col-xs-12 col-sm-6 pull-right"
    >
      <div
        className="input-group mb-2 mb-sm-0"
      >
        <div
          className="input-group-addon"
        >
          Search
        </div>
        <input
          className="form-control"
          id="search"
          onChange={[Function]}
          onKeyUp={[Function]}
          placeholder="Enter your search key words here!"
          type="text"
        />
      </div>
    </div>
    <div>
      <div
        className="col-xs-12"
      >
        <div
          className="Home jumbotron"
          id="search-bar"
        >
          <h3>
            No recipes matched your query.
          </h3>
        </div>
        <hr />
        <Recipes
          key="1"
          recipe_id={1}
          title="Lorem"
        >
          <div
            className="jumbotron"
          >
            <h3>
              Lorem
            </h3>
            <em>
              Added by 
              <span
                className="fa fa-user"
              />
               
               about 
              <span
                className="fa fa-calendar"
              />
               
              <Since
                live={true}
              />
               | Category: 
              <a
                className="btn btn-primary pull-right"
                href="/recipe/1"
                id="review"
              >
                Review
              </a>
            </em>
            <div>
              <h3>
                Ingredients
              </h3>
              <small />
              <h3>
                Steps
              </h3>
              <small />
              <br />
              <br />
              <div
                className="btn-group"
                id="btn-group"
              >
                <button
                  className="btn btn-default btn-xs"
                  type="button"
                >
                  <span
                    className="fa fa-comment-o"
                  />
                   Reviews:
                </button>
                <button
                  className="btn btn-default btn-xs"
                  type="button"
                >
                  <span
                    className="fa fa-thumbs-o-up"
                  />
                   Upvotes:
                </button>
              </div>
            </div>
          </div>
        </Recipes>
      </div>
      <div
        className="col-xs-11 col-sm-3 pull-right"
      >
        <ul
          className="pagination"
        >
          <li>
            <a
              className="page-link"
              id="prevPage"
              onClick={[Function]}
            >
              Previous
            </a>
          </li>
          <li>
            <a
              className="page-link"
              id="nextPage"
              onClick={[Function]}
            >
              Next
            </a>
            <a
              className="page-link"
            >
              Showing 
               of 
            </a>
          </li>
        </ul>
      </div>
    </div>
    <div />
  </div>
</Home>
`;
